rfcs/20211025-augru/README.md

```diff
@@ -0,0 +1,1089 @@
+Proposal for GRU with Attentional Update in DNNL
+================================================
+
+## 1. Introduction
+
+In addition to the regular GRU there is a notion of GRU with attentional update
+gate (AUGRU) that can help to model interest evolving in Click-Through Rate
+Prediction [[#1]][1].
+
+
+### 1.1. GRU
+
+GRU is defined by the following formula:
+
+![GRU-LBR](formula-gru-lbr.png)
+
+
+### 1.2. AUGRU
+
+AUGRU is defined by the following formula (the difference from the
+regular GRU is highlighted in red):
+
+![AUGRU-LBR](formula-augru-lbr.png)
+
+
+
+> Warning
+>
+> Attention is passed not as part of source layer as it is done in Neural Machine
+> Translation (NMT) models [[#3]][3], but as a separate memory which is
+> applied to intermediate results (to gate U). This difference makes it impossible
+> to apply attention on application side and pass the result to oneDNN RNN
+> primitive.
+
+## 2. API Changes
+
+Attention memory descriptor can be built from existing `rnn_desc_t` fields (memory
+descriptors, cell kind, direction, propagation kind, etc.). Taking into account
+this fact and to avoid breaking changes an attention memory descriptor will not
+be part of `dnnl_rnn_desc_t`. Instead a user should use query API to get
+attention shapes.
+
+ * Pros:
+   - RNN primitive descriptor can request the best layout for an attention memory
```

Comment:
This one is pretty scary actually. It has the potential to trigger spurious reorders on the user side.
Could you document what implementations are supposed to return and what FWK currently use (e.g. TNC, or NTC)? Is it ok to just use plain layouts for attention since it is involved only in binary operation with a matmul output.

Also, do you know how these attention vectors are generated? Do you expect those to be dense in memory or is there a chance they are strided or in different memories (e.g. 1 memory for each timestamp)?

Reply:
attention vectors are generated as a result of FCN network (matmul) as a single buffer which is then passed to AUGRU. They are in plain layout ([ai-matrix/utils.py](https://github.com/alibaba/ai-matrix/blob/f7e1d7788c03e01eea915ff2bcad10103dcb529e/macro_benchmark/DIEN_TF2/script/utils.py#L368))

Reply:
From the script you cited, it seems it is NTC by default (`time_major = false`), which typically differs from what we use for `_layer` tensors (`TNC`). We might need to check if applications keep this default so that we can determine what layout to return upon query.

Reply:
Formats of scores and src are aligned. So as long as we require score in the same format from our side we should be fine. I updated the RFC and added a diagram with AUGRU inputs. Please review.